File: android_java/app/src/main/java/com/fjun/androidjavaweatherlabapp/MainActivity.java
package com.fjun.androidjavaweatherlabapp;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

public class MainActivity extends AppCompatActivity implements MainActivityViewBinder {

    @Inject
    MainActivityPresenter mPresenter;

    private TextView mStatusTextView;
    private ImageView mImageView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mStatusTextView = findViewById(R.id.status);
        mImageView = findViewById(R.id.icon);

        // Setup presenter: Attach lifecycle
        getLifecycle().addObserver(mPresenter);
    }

    @Override
    public void setTemperature(String temperature) {
        mStatusTextView.setText(getString(R.string.temperature, temperature));
    }

    @Override
    public void showSomethingWentWrong(String error) {
        mStatusTextView.setText(getString(R.string.something_went_wrong, error));
    }

    @Override
    public void showWaitingForPosition() {
        mStatusTextView.setText(getString(R.string.waiting_for_position));
    }

    @Override
    public void showWaitingForWeather() {
        mStatusTextView.setText(getString(R.string.waiting_for_weather));
    }

    @Override
    public void setImageUrl(String imageUrl) {
        Picasso picasso = Picasso.get();
        picasso.load(imageUrl).into(mImageView);
    }

}

File: android_kotlin/app/src/main/java/com/fjun/android_kotlin/MainActivity.kt
package com.fjun.android_kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.activity_main.*
import org.koin.android.ext.android.inject

class MainActivity : AppCompatActivity(), MainActivityViewBinder {

    // Inject MyPresenter
    val presenter: MainActivityPresenter by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Setup presenter: Attach lifecycle
        lifecycle.addObserver(presenter)
    }

    override fun showSomethingWentWrong(error: String) {
        status.text = getString(R.string.something_went_wrong, error)
    }

    override fun showWaitingForPosition() {
        status.text = getString(R.string.waiting_for_position)
    }

    override fun showWaitingForWeather() {
        status.text = getString(R.string.waiting_for_weather)
    }

    override fun setTemperature(temperature: String) {
        status.text = getString(R.string.temperature, temperature)
    }

    override fun setImageUrl(imageUrl: String) {
        Glide.with(this).load(imageUrl).into(icon)
    }
}
